File: training-data/rust/knapsack_01.rs
fn knapsack(weights: &[usize], values: &[i32], capacity: usize) -> i32 {
    let n = weights.len();
    let mut dp = vec![vec![0; capacity + 1]; n + 1];
    for i in 1..=n {
        for w in 0..=capacity {
            dp[i][w] = dp[i - 1][w];
            if weights[i - 1] <= w {
                dp[i][w] = dp[i][w].max(dp[i - 1][w - weights[i - 1]] + values[i - 1]);
            }
        }
    }
    dp[n][capacity]
}

fn main() {
    println!("{}", knapsack(&[1, 3, 4, 5], &[1, 4, 5, 7], 7));
}

File: training-data/zig/knapsack_01.zig
const std = @import("std");

fn knapsack(weights: []const i32, values: []const i32, capacity: usize) i32 {
    var dp: [6]i32 = [_]i32{0} ** 6;
    for (weights, 0..) |w, i| {
        var cap = capacity;
        while (cap >= @as(usize, @intCast(w))) {
            const withItem = dp[cap - @as(usize, @intCast(w))] + values[i];
            if (withItem > dp[cap]) dp[cap] = withItem;
            if (cap == 0) break;
            cap -= 1;
        }
    }
    return dp[capacity];
}

pub fn main() void {
    const weights = [_]i32{ 2, 3, 4, 5 };
    const values = [_]i32{ 3, 4, 5, 6 };
    std.debug.print("{d}\n", .{knapsack(&weights, &values, 5)});
}
